### PR TITLE
Improve GUI message handling / gui::smartredraw

### PR DESCRIPTION
**Pull request by peak3d**
GUI optimizations (Message handling through lookup) and SmartRedraw (only Processes GUI if anything has changed).

## Description
1.) Use alreaedy existing gui-control-lookup for sending control messages.
2.) [Experimental]: If advancedsettings::gui:guismartredraw is set to true, the GUI Process call (and therefore the render call), wich is usually called on every vsync, is only done if anything is changed.

## Motivation and Context
Speed up GUI / Remove load if kodi idle / plays videos (factor ~10)

## How Has This Been Tested?
Win32 / Android / Esturay

## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

##Note
I'm pushing this PR quite early (regarding smartredraw) because @mkortstiege seems to play around with GUI.

### CODE REVIEW

File: xbmc/guilib/GUIWindowManager.cpp [hunk 1]
```diff
@@ -1148,7 +1148,12 @@ void CGUIWindowManager::AfterRender()
   for (const auto& window : activeDialogs)
   {
     if (window->IsDialogRunning())
+    {
       window->AfterRender();
+      // Dialog state can affect visibility states
+      if (pWindow && window->IsControlDirty())
```
Comment (Rechi) [inline xbmc/guilib/GUIWindowManager.cpp]: `class CGUIWindow` has no member named `IsControlDirty`

File: xbmc/cores/DataCacheCore.cpp [hunk 2]
```diff
@@ -235,3 +237,13 @@ bool CDataCacheCore::CDataCacheCore::IsSeeking()
 
   return m_stateInfo.m_stateSeeking;
 }
+
+bool CDataCacheCore::PlayStateChanged()
```
Comment (Rechi) [inline xbmc/cores/DataCacheCore.cpp]: definition in `DataCacheCore.h` is missing

File: xbmc/cores/DataCacheCore.cpp [hunk 3]
```diff
@@ -23,11 +23,12 @@
 #include "ServiceBroker.h"
 
 CDataCacheCore::CDataCacheCore()
+  :m_playStateChanged(false)
```
Comment (FernetMenta) [inline xbmc/cores/DataCacheCore.cpp]: init by assignment in header

File: xbmc/guilib/GUIWindowManager.cpp [hunk 4]
```diff
@@ -1174,6 +1175,9 @@ void CGUIWindowManager::FrameMove()
     m_deleteWindows.clear();
   }
 
+  if (CDataCacheCore::GetInstance().PlayStateChanged())
```
Comment (FernetMenta) [inline xbmc/guilib/GUIWindowManager.cpp]: CApplicationPlayer::FrameMove is a much better place to check this. That method also knows if a player is active

File: xbmc/cores/DataCacheCore.cpp (hunk 2 above)
Comment (FernetMenta) [inline xbmc/cores/DataCacheCore.cpp]: please rename to PlayerStateChenged

File: xbmc/cores/DataCacheCore.cpp [hunk 5]
```diff
@@ -227,6 +227,7 @@ void CDataCacheCore::SetStateSeeking(bool active)
   CSingleLock lock(m_stateSection);
 
   m_stateInfo.m_stateSeeking = active;
+  m_playStateChanged = true;
```
Comment (Rechi) [inline xbmc/cores/DataCacheCore.cpp]: `m_playStateChanged` was not declared in this scope

Comment (Rechi) [inline xbmc/cores/DataCacheCore.cpp]: `m_playerStateChanged`

**Comment by peak3d** [comment]
jenkins build this please

**Comment by peak3d** [comment]
Any objections @FernetMenta ? Its approved from popcornmix (the last commit) and the others are running quite some days on my home equipment.

File: xbmc/Application.cpp [hunk 6]
```diff
@@ -33,6 +33,7 @@
 #include "Util.h"
 #include "URL.h"
 #include "guilib/TextureManager.h"
+#include "cores/DataCacheCore.h"
```
Comment (FernetMenta) [inline xbmc/Application.cpp]: left over from previous attempt

File: xbmc/Application.cpp [hunk 7]
```diff
@@ -5232,6 +5235,12 @@ void CApplication::CloseNetworkShares()
   }
 }
 
+void CApplication::SendGUIMessage(int dwMsg, int senderID/* = 0*/, int controlID/* = 0*/, int param1/* = 0*/, int param2/* = 0*/)
```
Comment (FernetMenta) [inline xbmc/Application.cpp]: https://github.com/xbmc/xbmc/blob/master/xbmc/messaging/ApplicationMessenger.cpp#L272

please elaborate why you create an additionsl SendGUIMessage

File: xbmc/dialogs/GUIDialogSeekBar.cpp [hunk 8]
```diff
@@ -30,6 +30,7 @@
 
 CGUIDialogSeekBar::CGUIDialogSeekBar(void)
   : CGUIDialog(WINDOW_DIALOG_SEEK_BAR, "DialogSeekBar.xml", DialogModalityType::MODELESS)
+  , m_lastPercent(~0)
```
Comment (FernetMenta) [inline xbmc/dialogs/GUIDialogSeekBar.cpp]: please init by assignment in header

**Comment by peak3d** [comment]
jenkins build this please

**Comment by peak3d** [comment]
Because you build the wrong ref: jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by HitcherUK** [comment]
This could be the cause of problems with scrollbars and dialog grouplist buttons not fading when not focused. https://forum.kodi.tv/showthread.php?tid=316560

**Comment by HitcherUK** [comment]
@peak3d I can confirm that this is the cause of the grouplist problems mentioned [here](https://forum.kodi.tv/showthread.php?tid=316560&pid=2604968#pid2604968) and reported [here](https://trac.kodi.tv/ticket/17504). This was tested using a [build](http://mirrors.kodi.tv/test-builds/windows/win32/KodiSetup-20170605-a16e0da-PR12213exc-x86.exe) before this commit and a [build ](http://mirrors.kodi.tv/test-builds/windows/win32/KodiSetup-20170606-40a1a0c513-PR12213inc-x86.exe)after.

Thanks.

**Comment by pkerling** [comment]
@peak3d I was seeing another (i.e. not yet reported as far as I know) strange behavior in Kodi for some time and finally got around to bisecting it. Commit 40a1a0c5 in this PR introduces it:
You can see it by opening e.g. the system settings dialog and then move the mouse around at the upper left edge where it says "Settings / System". The currently selected category item will randomly flicker (it did not before). It also flickers sometimes when you move the mouse over the individual settings on the right side.

**Comment by peak3d** [comment]
thanks @pkerling for the report, yes, I have noticed this one as well, but hoped that nobody else will see it (joking). The selection / focus things are a little bit messed up in GUI, I have one more issue and will focus on it next week (unless you already have the reason for it (???) :-)

**Comment by pkerling** [comment]
@peak3d No I didn't look further into it, only bisected

File: xbmc/Application.cpp [hunk 9]
```diff
@@ -4594,6 +4594,8 @@ void CApplication::ProcessSlow()
 
   m_ServiceManager->GetActiveAE().GarbageCollect();
 
+  g_windowManager.SendMessage(GUI_MSG_REFRESH_TIMER,0,0);
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/Application.cpp]: no idea what this is for, pvr does not use  this message (afaik)

[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/Application.cpp]: ok, I'll move it to FrameMove

[the remark this answers is not in the document]
Comment (peak3d) [reply xbmc/Application.cpp]: This message triggers a "change detection" in GUI in smartredraw mode.
It should NOT be done on every FrameMove, it should be done on reasonable time.

Example is the GUI clock wich forces GUI actions on each frame, 500ms between the refresh (ProcessSlow) is quite fine, so FrameMove is not the best option

[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/Application.cpp]: already done

**Comment by JimmyS83** [comment]
_If advancedsettings::gui:guismartredraw is set to true, the GUI Process call (and therefore the render call), wich is usually called on every vsync, is only done if anything is changed._

Does it means, when black screensaver is active, then there is no render at all and CPU load is reduced?

**Comment by peak3d** [comment]
yes, schould be. There is still load because of other threads (network e.g.), but CPU load is in general very low.

**Comment by JimmyS83** [comment]
> 
> 
> yes, schould be. There is still load because of other threads (network e.g.), but CPU load is in general very low.

Just start testing this feature.. I think CPU is still [higher than its needed](https://forum.kodi.tv/showthread.php?tid=338742&pid=2803158#pid2803158) - compare with Jarvis release, and I have the same network / configuration / addons...  But at least I see some CPU improvement with smart redraw (I didnt do any actual user testing, so cant comment more now).

**Comment by peak3d** [comment]
There are still some GUI elements which don't draw / update correctly.
If benefit is not high for you (for me on low powered Android devices it is), than its better to disable.

All this depends on what addons - especially service addons - do in background. I have no influence on this.

**Comment by JimmyS83** [comment]
Any lowering CPU is benefit for me. The Rpi / C2 boxes running 24/7, and actually get used 1-4h per day.

I am almost sure that no addons causes this difference, because I use just a few of them, and all of them (and also configuration of services like Samba, CEC, etc. remains exactly same) already used before on Rpi2+Jarvis with 1-3% CPU..

**Comment by JimmyS83** [comment]
Is this good platform to report bugs? I didnt find forum thread, so if its not bother you here:
- Confluence skin, little window showing library update progress doesnt dissappears, after library update.

**Comment by MartijnKaijser** [comment]
@JimmyS83 no it's not. Use the correct way and create a github ticket with enough info. Do not start asking questions in a very old PR

**Comment by zaphod24** [comment]
What is the recommended setting for dirty regions when smartredraw is set to true? There seem to be multiple opinions so I figured I’d ask here. Unless something has changed in Leia, the default for dirty regions is still 3. Should it be set to zero to disable dirty regions? Should only smartredraw=true be specified leaving dirty regions defaulted to 3?

For reference I’m running Leia on a c2

**Pull request merged**